src/app/clusters/testing/ClusterTesting.h

```diff
@@ -0,0 +1,97 @@
+// For example, in the case of ReadAttribute(), the call to app::DataModel::Decode() for the given type T should be supported.
+// Otherwise, that template instantiation will fail.
+// This helper class can also be used with attributes that are lists or strings because it maintains the read/write memory.
+// This is why ReadOperation and WriteOperation are data members.
+//
+// Example of usage:
+//
+// FixedLabelCluster fixedLabel(kRootEndpointId);
+//
+// ClusterTester tester(fixedLabel);
+// uint32_t features{};
+// ASSERT_EQ(tester.ReadAttribute(FeatureMap::Id, features), CHIP_NO_ERROR);
+//
+// DataModel::DecodableList<Structs::LabelStruct::Type> labelList;
+// ASSERT_EQ(tester.ReadAttribute(LabelList::Id, labelList), CHIP_NO_ERROR);
+// auto it = labelList.begin();
+// while (it.Next())
+// {
+//     ASSERT_GT(it.GetValue().label.size(), 0u);
+// }
+class ClusterTester
+{
+public:
+    ClusterTester(app::ServerClusterInterface & cluster) : mCluster(cluster) {}
+
+    template <typename T>
+    CHIP_ERROR ReadAttribute(AttributeId attr, T & value)
+    {
+        const auto & paths = mCluster.GetPaths();
+        // This must be a size-1 span
+        VerifyOrReturnError(paths.size() == 1u, CHIP_ERROR_INCORRECT_STATE);
+        app::ConcreteAttributePath attributePath(paths[0].mEndpointId, paths[0].mClusterId, attr);
+        mReadOperation = std::make_unique<app::Testing::ReadOperation>(attributePath);
+
+        std::unique_ptr<app::AttributeValueEncoder> encoder = mReadOperation->StartEncoding();
+        ReturnErrorOnFailure(mCluster.ReadAttribute(mReadOperation->GetRequest(), *encoder).GetUnderlyingError());
+        ReturnErrorOnFailure(mReadOperation->FinishEncoding());
+
+        std::vector<app::Testing::DecodedAttributeData> attributeData;
+        ReturnErrorOnFailure(mReadOperation->GetEncodedIBs().Decode(attributeData));
+        VerifyOrReturnError(attributeData.size() == 1u, CHIP_ERROR_INCORRECT_STATE);
+
+        return app::DataModel::Decode(attributeData[0].dataReader, value);
+    }
+
+    template <typename T>
+    CHIP_ERROR WriteAttribute(AttributeId attr, const T & value)
+    {
+        const auto & paths = mCluster.GetPaths();
+        // This must be a size-1 span
+        VerifyOrReturnError(paths.size() == 1u, CHIP_ERROR_INCORRECT_STATE);
+        app::ConcreteAttributePath attributePath(paths[0].mEndpointId, paths[0].mClusterId, attr);
+        mWriteOperation                    = std::make_unique<app::Testing::WriteOperation>(attributePath);
+        app::AttributeValueDecoder decoder = mWriteOperation->DecoderFor(value);
+        return mCluster.WriteAttribute(mWriteOperation->GetRequest(), decoder).GetUnderlyingError();
+    };
+
+private:
+    app::ServerClusterInterface & mCluster;
+    std::unique_ptr<app::Testing::ReadOperation> mReadOperation;
+    std::unique_ptr<app::Testing::WriteOperation> mWriteOperation;
```

Comment:
For reads we need to return the value so we need a member. But for writes we do not need it: write is a one-shot operation. Lets drop this member and just place the operation inline.